internal/project/validation.go

```diff
@@ -1,18 +1,47 @@
 package project
 
 import (
+	"context"
 	"fmt"
+	"regexp"
+	"strings"
 
+	"github.com/razorpay/metro/pkg/logger"
 	metrov1 "github.com/razorpay/metro/rpc/proto/v1"
 )
 
-// Validate an incoming proto request
-func Validate(req *metrov1.Project) error {
+// GetValidatedModel validates an incoming proto request and returns a project model
+func GetValidatedModel(ctx context.Context, req *metrov1.Project) (*Model, error) {
 	if req.Name == "" {
-		return fmt.Errorf("name cannot be empty")
+		return nil, fmt.Errorf("name cannot be empty")
 	}
 	if req.ProjectId == "" {
-		return fmt.Errorf("projectId cannot be empty")
+		return nil, fmt.Errorf("projectId cannot be empty")
 	}
-	return nil
+	if isValidProjectID(ctx, req.ProjectId) {
+		return fromProto(req), nil
+	}
+	return nil, fmt.Errorf("not a valid projectID")
+}
+
+// fromProto creates and returns a Model from proto message
+func fromProto(proto *metrov1.Project) *Model {
+	m := &Model{}
+	m.Name = proto.GetName()
+	m.ProjectID = proto.GetProjectId()
+	m.Labels = proto.GetLabels()
+	return m
+}
+
+func isValidProjectID(ctx context.Context, projectID string) bool {
+	if strings.HasSuffix(projectID, "-") {
+		logger.Ctx(ctx).Error("projectID cannot end with a trailing hyphen")
+		return false
+	}
+	r, err := regexp.Compile("([a-z][a-z0-9-]{5,29})$")
```

Comment:
```suggestion
	var idRegex, _ := regexp.Compile("([a-z][a-z0-9-]{5,29})$")
```
We can export this to a var outside the function and ignore the error since this would be a one-time expr setup. Will avoid recompilation on every function call.